### PR TITLE
dialects: (builtin): remove IndexType support for DenseIntOrFPElementsAttr

### PR DESCRIPTION
**Pull request by jorendumoulin**
This removes support for creating dense attributes with `IndexType` values.

### CODE REVIEW

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -1801,61 +1783,28 @@ def create_dense_float(
...
-    def from_list(
-        type: (
-            RankedStructure[AnyFloat | IntegerType | IndexType]
-            | RankedStructure[AnyFloat]
-            | RankedStructure[IntegerType]
-            | RankedStructure[IndexType]
-        ),
-        data: Sequence[int | float] | Sequence[AnyFloatAttr],
-    ) -> DenseIntOrFPElementsAttr: ...
-
```
Comment (jorendumoulin) [inline xdsl/dialects/builtin.py:1831]: I think these overloads achieve nothing?

Comment (superlopuh) [reply xdsl/dialects/builtin.py:1831]: It's very possible that they've drifted

File: xdsl/parser/attribute_parser.py [hunk 2]
```diff
@@ -864,6 +858,7 @@ def parse_dense_int_or_fp_elements_attr(
                 assert len(data_values) == 1, "Fatal error in parser"
                 data_values *= type_num_values
 
+        assert type
```
Comment (superlopuh) [inline xdsl/parser/attribute_parser.py:861]: would be good to remove before merging

**Comment by superlopuh** [comment]
We should just replace the failing tests' types to not use `index`, which would also fail in MLIR IIRC

**Review by superlopuh — APPROVED** [review]

**Comment by superlopuh** [comment]
Joren, I'm sorry, I just played locally with it and it looks like MLIR does support index here... I'm not sure how it actually works, though, because the size of index is not known in general. It would be good to replicate the behaviour of MLIR, they probably do what you initially suggested and just use a long long in until it's lowered to a concrete bitwidth.

**Comment by jorendumoulin** [comment]
Hold up!

I don't know why I hadn't checked this before, but this seems to be perfectly valid in MLIR:

```
module {
  memref.global "public" @g : memref<5xindex> = dense<[0, 1, 2, 3, 4]>
}
```

I'm not sure anymore we want to remove support for this?

**Comment by jorendumoulin** [comment]
> Joren, I'm sorry, I just played locally with it and it looks like MLIR does support index here... I'm not sure how it actually works, though, because the size of index is not known in general. It would be good to replicate the behaviour of MLIR, they probably do what you initially suggested and just use a long long in until it's lowered to a concrete bitwidth.

Yes I just noticed as well! So probably for this PR just close it entirely?

**Comment by math-fehr** [comment]
A bit of information, `index` is always less than 64 bits, so they probably store values on 64 bits then trunc them later on.

**Pull request closed without merging**